File: trevrpc-js/src/framing.js
import { FrameTooLargeError, invalidArgument, unavailable } from "./status.js";
import { RpcStreamFrame, RpcStreamFrameKind } from "./wire.js";

export const DefaultMaxFrameSize = 4 * 1024 * 1024;

const FrameHeaderLength = 4;
const StreamFrameBodyTag = (4 << 3) | 2;
const EmptyBytes = new Uint8Array(0);
const utf8Decoder = new TextDecoder("utf-8", { fatal: true });

/** Encodes a protobuf message body. */
export function marshalMessage(messageType, message) {
  const prepared = prepareMessage(messageType, message);
  return messageType.encode(prepared).finish();
}

/** Decodes a protobuf message body. */
export function unmarshalMessage(messageType, body) {
  try {
    return messageType.decode(body);
  } catch (error) {
    throw invalidArgument(`failed to decode protobuf message: ${error?.message ?? String(error)}`);
  }
}

/** Encodes a protobuf message into a length-prefixed TrevRPC frame. */
export function encodeFrame(messageType, message, maxFrameSize = DefaultMaxFrameSize) {
  const body = marshalMessage(messageType, message);
  if (body.byteLength > maxFrameSize) {
    throw new FrameTooLargeError(body.byteLength, maxFrameSize);
  }

  const frame = new Uint8Array(FrameHeaderLength + body.byteLength);
  const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);
  view.setUint32(0, body.byteLength, false);
  frame.set(body, FrameHeaderLength);
  return frame;
}

/** Encodes one streaming message frame carrying an already-encoded protobuf body. */
export function encodeMessageStreamFrame(body, maxFrameSize = DefaultMaxFrameSize) {
  return encodeMessageStreamFrames([body], maxFrameSize);
}

/** Encodes multiple streaming message frames into one contiguous write buffer. */
export function encodeMessageStreamFrames(
  bodies,
  maxFrameSize = DefaultMaxFrameSize,
  start = 0,
  end = undefined,
) {
  const range = bodyRange(bodies, start, end);
  let totalLength = 0;
  const byteBodies = Array.from({ length: range.end - range.start });
  for (let index = range.start; index < range.end; index += 1) {
    const body = byteView(range.bodies[index]);
    byteBodies[index - range.start] = body;
    const bodyLength = messageStreamFrameBodyLength(body.byteLength);
    if (bodyLength > maxFrameSize) {
      throw new FrameTooLargeError(bodyLength, maxFrameSize);
    }
    totalLength += FrameHeaderLength + bodyLength;
  }

  const frames = new Uint8Array(totalLength);
  const view = new DataView(frames.buffer, frames.byteOffset, frames.byteLength);
  let offset = 0;
  for (const body of byteBodies) {
    const bodyLength = messageStreamFrameBodyLength(body.byteLength);
    view.setUint32(offset, bodyLength, false);
    offset += FrameHeaderLength;
    if (body.byteLength > 0) {
      frames[offset++] = StreamFrameBodyTag;
      offset = appendVarint(frames, offset, body.byteLength);
      frames.set(body, offset);
      offset += body.byteLength;
    }
  }

  return frames;
}

/** Encodes each streaming message frame separately for vectored writers. */
export function encodeMessageStreamFrameParts(
  bodies,
  maxFrameSize = DefaultMaxFrameSize,
  start = 0,
  end = undefined,
) {
  const range = bodyRange(bodies, start, end);
  const frames = Array.from({ length: range.end - range.start });
  for (let index = range.start; index < range.end; index += 1) {
    frames[index - range.start] = encodeMessageStreamFrame(range.bodies[index], maxFrameSize);
  }
  return frames;
}

/** Decodes a protobuf message from a TrevRPC frame body. */
export function decodeFrame(messageType, body) {
  return unmarshalMessage(messageType, body);
}

/** Decodes a streaming RPC frame body using a fast path for common message frames. */
export function decodeStreamFrameBody(body) {
  const bytes = byteView(body);
  if (bytes.byteLength === 0) {
    return streamFrameMessage(EmptyBytes);
  }

  const cursor = { offset: 0 };
  let kind = RpcStreamFrameKind.Message;
  let status = 0;
  let message = "";
  let frameBody = new Uint8Array(0);

  while (cursor.offset < bytes.byteLength) {
    const tag = consumeVarint(bytes, cursor, "invalid stream frame field tag");
    if (tag === 0) {
      throw invalidArgument("invalid stream frame field tag");
    }

    const field = Math.floor(tag / 8);
    const wireType = tag % 8;
    switch (field) {
      case 1:
        requireWireType(wireType, 0, "invalid stream frame kind wire type");
        kind = consumeVarint(bytes, cursor, "truncated stream frame kind");
        break;
      case 2:
        requireWireType(wireType, 0, "invalid stream frame status wire type");
        status = consumeVarint(bytes, cursor, "truncated stream frame status");
        break;
      case 3:
        requireWireType(wireType, 2, "invalid stream frame message wire type");
        message = decodeUtf8(consumeLengthDelimited(bytes, cursor, "invalid stream frame message"));
        break;
      case 4:
        requireWireType(wireType, 2, "invalid stream frame body wire type");
        frameBody = consumeLengthDelimited(bytes, cursor, "invalid stream frame body");
        break;
      case 5:
        return decodeFrame(RpcStreamFrame, bytes);
      default:
        skipProtoField(bytes, cursor, wireType);
        break;
    }
  }

  if (kind !== RpcStreamFrameKind.Message && kind !== RpcStreamFrameKind.Status) {
    throw invalidArgument("stream frame contained an unknown frame kind");
  }

  return { kind, status, message, body: frameBody, metadata: {} };
}

/** Writes one length-prefixed protobuf frame. */
export async function writeFrame(writer, messageType, message, maxFrameSize = DefaultMaxFrameSize) {
  await writer.write(encodeFrame(messageType, message, maxFrameSize));
}

/** Writes multiple streaming message frames in one write. */
export async function writeMessageStreamFrames(
  writer,
  bodies,
  maxFrameSize = DefaultMaxFrameSize,
  start = 0,
  end = undefined,
) {
  if (typeof writer.writev === "function") {
    await writer.writev(encodeMessageStreamFrameParts(bodies, maxFrameSize, start, end));
    return;
  }

  await writer.write(encodeMessageStreamFrames(bodies, maxFrameSize, start, end));
}

/** Reads length-prefixed protobuf frames from a byte stream reader. */
export class FrameReader {
  /** Creates a frame reader over a stream reader. */
  constructor(reader) {
    this.reader = reader;
    this.chunks = [];
    this.chunkHead = 0;
    this.buffered = 0;
  }

  /** Reads and decodes one frame. */
  async readFrame(messageType, maxFrameSize = DefaultMaxFrameSize) {
    const header = await this.readExact(FrameHeaderLength, false);
    const length = frameBodyLength(header, maxFrameSize);
    const body = await this.readExact(length, false);
    return decodeFrame(messageType, body);
  }

  /** Reads and decodes one frame, or returns null when already at EOF. */
  async readFrameOrEOF(messageType, maxFrameSize = DefaultMaxFrameSize) {
    const header = await this.readExact(FrameHeaderLength, true);
    if (header == null) {
      return null;
    }

    const length = frameBodyLength(header, maxFrameSize);
    const body = await this.readExact(length, false);
    return decodeFrame(messageType, body);
  }

  /** Reads and decodes one streaming RPC frame, or returns null when already at EOF. */
  async readStreamFrameOrEOF(maxFrameSize = DefaultMaxFrameSize) {
    const header = await this.readExact(FrameHeaderLength, true);
    if (header == null) {
      return null;
    }

    const length = frameBodyLength(header, maxFrameSize);
    const body = await this.readExact(length, false);
    return decodeStreamFrameBody(body);
  }

  /** Reads one streaming RPC frame, then drains complete frames already buffered. */
  async readStreamFrameBatchOrEOF(maxBatch = 32, maxFrameSize = DefaultMaxFrameSize) {
    const first = await this.readStreamFrameOrEOF(maxFrameSize);
    if (first == null) {
      return null;
    }

    const limit = Math.max(1, Math.floor(maxBatch));
    const frames = [first];
    while (frames.length < limit) {
      const body = this.tryReadFrameBody(maxFrameSize);
      if (body === undefined) {
        break;
      }
      frames.push(decodeStreamFrameBody(body));
    }
    return frames;
  }

  /** Reads stream message body batches, with at most one terminal status. */
  async readStreamMessageBodyBatchOrEOF(maxBatch = 32, maxFrameSize = DefaultMaxFrameSize) {
    const body = await this.readRawFrameBodyOrEOF(maxFrameSize);
    if (body == null) {
      return null;
    }

    const limit = Math.max(1, Math.floor(maxBatch));
    const result = { bodies: [], status: null };
    pushStreamBodyOrStatus(result, body);
    while (result.status == null && result.bodies.length < limit) {
      const nextBody = this.tryReadFrameBody(maxFrameSize);
      if (nextBody === undefined) {
        break;
      }
      pushStreamBodyOrStatus(result, nextBody);
    }

    return result;
  }

  /** Reads one raw frame body, or returns null when already at EOF. */
  async readRawFrameBodyOrEOF(maxFrameSize = DefaultMaxFrameSize) {
    const header = await this.readExact(FrameHeaderLength, true);
    if (header == null) {
      return null;
    }

    return this.readExact(frameBodyLength(header, maxFrameSize), false);
  }

  /** Reads a complete buffered frame body without waiting, or undefined. */
  tryReadFrameBody(maxFrameSize = DefaultMaxFrameSize) {
    if (this.buffered < FrameHeaderLength) {
      return undefined;
    }

    const header = this.peek(FrameHeaderLength);
    const length = frameBodyLength(header, maxFrameSize);
    if (this.buffered < FrameHeaderLength + length) {
      return undefined;
    }

    this.consume(FrameHeaderLength);
    return this.consume(length);
  }

  /** Reads exactly size bytes from the underlying reader. */
  async readExact(size, allowEofAtStart) {
    if (size === 0) {
      return EmptyBytes;
    }

    while (this.buffered < size) {
      const { value, done } = await this.reader.read();
      if (done) {
        if (allowEofAtStart && this.buffered === 0) {
          return null;
        }

        throw unavailable("transport unavailable: stream ended in the middle of a frame");
      }

      if (value == null || value.byteLength === 0) {
        continue;
      }

      const bytes = value instanceof Uint8Array ? value : new Uint8Array(value);
      this.chunks.push(bytes);
      this.buffered += bytes.byteLength;
    }

    return this.consume(size);
  }

  /** Removes size bytes from the buffered data. */
  consume(size) {
    if (size === 0) {
      return EmptyBytes;
    }

    const first = this.chunks[this.chunkHead];
    if (first != null && first.byteLength >= size) {
      const result = first.subarray(0, size);
      if (first.byteLength === size) {
        this.chunkHead += 1;
        this.compactChunks();
      } else {
        this.chunks[this.chunkHead] = first.subarray(size);
      }
      this.buffered -= size;
      return result;
    }

    const result = new Uint8Array(size);
    let offset = 0;

    while (offset < size) {
      const chunk = this.chunks[this.chunkHead];
      const needed = size - offset;
      if (chunk.byteLength <= needed) {
        result.set(chunk, offset);
        offset += chunk.byteLength;
        this.chunkHead += 1;
      } else {
        result.set(chunk.subarray(0, needed), offset);
        this.chunks[this.chunkHead] = chunk.subarray(needed);
        offset += needed;
      }
    }

    this.buffered -= size;
    this.compactChunks();
    return result;
  }

  /** Copies size bytes from the buffered data without removing them. */
  peek(size) {
    if (size === 0) {
      return EmptyBytes;
    }

    const first = this.chunks[this.chunkHead];
    if (first != null && first.byteLength >= size) {
      return first.subarray(0, size);
    }

    const result = new Uint8Array(size);
    let offset = 0;

    for (let index = this.chunkHead; index < this.chunks.length; index += 1) {
      const chunk = this.chunks[index];
      if (offset >= size) {
        break;
      }
      const count = Math.min(chunk.byteLength, size - offset);
      result.set(chunk.subarray(0, count), offset);
      offset += count;
    }

    return result;
  }

  compactChunks() {
    if (this.chunkHead === 0) {
      return;
    }
    if (this.chunkHead === this.chunks.length) {
      this.chunks.length = 0;
      this.chunkHead = 0;
      return;
    }
    if (this.chunkHead >= 32 && this.chunkHead * 2 >= this.chunks.length) {
      this.chunks.splice(0, this.chunkHead);
      this.chunkHead = 0;
    }
  }
}

function pushStreamBodyOrStatus(result, body) {
  const messageBody = tryDecodeStreamMessageBody(body);
  if (messageBody !== undefined) {
    result.bodies.push(messageBody);
    return;
  }

  const frame = decodeStreamFrameBody(body);
  switch (frame.kind) {
    case RpcStreamFrameKind.Message:
      result.bodies.push(frame.body ?? EmptyBytes);
      break;
    case RpcStreamFrameKind.Status:
      result.status = frame;
      break;
    default:
      throw invalidArgument("stream frame contained an unknown frame kind");
  }
}

function tryDecodeStreamMessageBody(body) {
  const bytes = byteView(body);
  if (bytes.byteLength === 0) {
    return EmptyBytes;
  }

  if (bytes[0] !== StreamFrameBodyTag) {
    return undefined;
  }

  const cursor = { offset: 1 };
  const length = consumeVarint(bytes, cursor, "invalid stream frame body");
  const end = cursor.offset + length;
  if (end !== bytes.byteLength) {
    return undefined;
  }
  return bytes.subarray(cursor.offset, end);
}

/** Decodes and validates the body length stored in a TrevRPC frame header. */
export function frameBodyLength(header, maxFrameSize = DefaultMaxFrameSize) {
  const view = new DataView(header.buffer, header.byteOffset, header.byteLength);
  const length = view.getUint32(0, false);
  if (length > maxFrameSize) {
    throw new FrameTooLargeError(length, maxFrameSize);
  }

  return length;
}

function prepareMessage(messageType, message) {
  const prepared =
    typeof messageType.fromObject === "function" ? messageType.fromObject(message) : message;
  const error = typeof messageType.verify === "function" ? messageType.verify(prepared) : null;
  if (error != null) {
    throw invalidArgument(`failed to encode protobuf message: ${error}`);
  }

  return prepared;
}

function streamFrameMessage(body) {
  return {
    kind: RpcStreamFrameKind.Message,
    status: 0,
    message: "",
    body,
    metadata: {},
  };
}

function byteView(value) {
  if (value instanceof Uint8Array) {
    return value;
  }
  if (ArrayBuffer.isView(value)) {
    return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
  }
  return new Uint8Array(value ?? 0);
}

function bodyRange(bodies, start, end) {
  const source = Array.isArray(bodies) ? bodies : Array.from(bodies);
  const from = clampRangeIndex(start, source.length);
  const to = Math.max(from, clampRangeIndex(end ?? source.length, source.length));
  return { bodies: source, start: from, end: to };
}

function clampRangeIndex(value, length) {
  if (!Number.isFinite(value)) {
    return length;
  }
  return Math.min(length, Math.max(0, Math.floor(value)));
}

function messageStreamFrameBodyLength(bodyLength) {
  return bodyLength === 0 ? 0 : 1 + varintLength(bodyLength) + bodyLength;
}

function appendVarint(output, offset, value) {
  let remaining = value;
  while (remaining >= 0x80) {
    output[offset++] = (remaining % 0x80) | 0x80;
    remaining = Math.floor(remaining / 0x80);
  }
  output[offset++] = remaining;
  return offset;
}

function varintLength(value) {
  let remaining = value;
  let length = 1;
  while (remaining >= 0x80) {
    remaining = Math.floor(remaining / 0x80);
    length += 1;
  }
  return length;
}

function consumeVarint(data, cursor, message) {
  let value = 0;
  let multiplier = 1;
  for (let index = 0; index < 10; index += 1) {
    if (cursor.offset >= data.byteLength) {
      throw invalidArgument(message);
    }
    const byte = data[cursor.offset++];
    value += (byte & 0x7f) * multiplier;
    if (byte < 0x80) {
      return value;
    }
    multiplier *= 0x80;
  }
  throw invalidArgument("stream frame varint exceeded 64 bits");
}

function requireWireType(actual, expected, message) {
  if (actual !== expected) {
    throw invalidArgument(message);
  }
}

function consumeLengthDelimited(data, cursor, message) {
  const length = consumeVarint(data, cursor, message);
  const end = cursor.offset + length;
  if (!Number.isSafeInteger(end) || end > data.byteLength) {
    throw invalidArgument(message);
  }
  const value = data.subarray(cursor.offset, end);
  cursor.offset = end;
  return value;
}

function skipProtoField(data, cursor, wireType) {
  switch (wireType) {
    case 0:
      consumeVarint(data, cursor, "truncated stream frame varint");
      return;
    case 1:
      skipFixed(data, cursor, 8);
      return;
    case 2:
      consumeLengthDelimited(data, cursor, "truncated stream frame field");
      return;
    case 5:
      skipFixed(data, cursor, 4);
      return;
    default:
      throw invalidArgument("stream frame contained an unsupported wire type");
  }
}

function skipFixed(data, cursor, length) {
  const end = cursor.offset + length;
  if (end > data.byteLength) {
    throw invalidArgument("truncated stream frame fixed field");
  }
  cursor.offset = end;
}

function decodeUtf8(bytes) {
  try {
    return utf8Decoder.decode(bytes);
  } catch (error) {
    throw invalidArgument(`stream frame message was not valid UTF-8: ${error.message}`);
  }
}

File: trevrpc-rust/src/framing.rs
use prost::Message;

use crate::{Code, Error, Metadata, Result, RpcStreamFrame, RpcStreamFrameKind, Status};

const FRAME_HEADER_LEN: usize = 4;
const STREAM_FRAME_BODY_TAG: u8 = 4 << 3 | 2;

pub const DEFAULT_MAX_FRAME_SIZE: usize = 4 * 1024 * 1024;

/// Encodes a protobuf message with the default `TrevRPC` frame size limit.
pub fn encode_frame<M>(message: &M) -> Result<Vec<u8>>
where
    M: Message,
{
    encode_frame_with_max(message, DEFAULT_MAX_FRAME_SIZE)
}

/// Encodes a protobuf message into a length-prefixed `TrevRPC` frame.
pub fn encode_frame_with_max<M>(message: &M, max_frame_size: usize) -> Result<Vec<u8>>
where
    M: Message,
{
    let encoded_len = message.encoded_len();

    if encoded_len > max_frame_size {
        return Err(Error::FrameTooLarge {
            len: encoded_len,
            max: max_frame_size,
        });
    }

    let frame_len = u32::try_from(encoded_len).map_err(|_| Error::FrameTooLarge {
        len: encoded_len,
        max: max_frame_size,
    })?;

    let mut frame = Vec::with_capacity(FRAME_HEADER_LEN + encoded_len);
    frame.extend_from_slice(&frame_len.to_be_bytes());
    message.encode(&mut frame)?;

    Ok(frame)
}

/// Encodes one stream message frame carrying an already-encoded protobuf body.
pub fn encode_message_stream_frame(body: &[u8], max_frame_size: usize) -> Result<Vec<u8>> {
    let body_len = message_stream_frame_body_len(body.len());
    check_frame_body_len(body_len, max_frame_size)?;

    let mut frame = Vec::with_capacity(FRAME_HEADER_LEN + body_len);
    frame.extend_from_slice(&frame_len(body_len, max_frame_size)?.to_be_bytes());
    append_message_stream_frame_body(&mut frame, body);

    Ok(frame)
}

/// Encodes multiple stream message frames into one contiguous write buffer.
pub fn encode_message_stream_frames(bodies: &[Vec<u8>], max_frame_size: usize) -> Result<Vec<u8>> {
    let mut total_len = 0_usize;
    for body in bodies {
        let body_len = message_stream_frame_body_len(body.len());
        check_frame_body_len(body_len, max_frame_size)?;
        total_len = total_len.saturating_add(FRAME_HEADER_LEN + body_len);
    }

    let mut frames = Vec::with_capacity(total_len);
    for body in bodies {
        let body_len = message_stream_frame_body_len(body.len());
        frames.extend_from_slice(&frame_len(body_len, max_frame_size)?.to_be_bytes());
        append_message_stream_frame_body(&mut frames, body);
    }

    Ok(frames)
}

/// Decodes a streaming RPC frame body using a fast path for common message frames.
pub fn decode_stream_frame_body(body: &[u8]) -> Result<RpcStreamFrame> {
    if body.is_empty() {
        return Ok(RpcStreamFrame::message(Vec::new()));
    }

    let mut offset = 0;
    let mut kind = RpcStreamFrameKind::Message as i32;
    let mut status = Code::Ok.as_u32();
    let mut message = String::new();
    let mut frame_body = Vec::new();
    let metadata = Metadata::new();

    while offset < body.len() {
        let tag = consume_varint(body, &mut offset)?;
        if tag == 0 {
            return Err(invalid_stream_frame("invalid stream frame field tag"));
        }

        let field = tag >> 3;
        let wire_type = tag & 0x7;
        match field {
            1 => {
                require_wire_type(wire_type, 0, "invalid stream frame kind wire type")?;
                kind = i32::try_from(consume_varint(body, &mut offset)?).map_err(|_| {
                    invalid_stream_frame("stream frame kind exceeded supported range")
                })?;
            }
            2 => {
                require_wire_type(wire_type, 0, "invalid stream frame status wire type")?;
                status = u32::try_from(consume_varint(body, &mut offset)?).map_err(|_| {
                    invalid_stream_frame("stream frame status exceeded supported range")
                })?;
            }
            3 => {
                require_wire_type(wire_type, 2, "invalid stream frame message wire type")?;
                let value = consume_length_delimited(body, &mut offset)?;
                message = String::from_utf8(value.to_vec()).map_err(|_| {
                    invalid_stream_frame("stream frame message was not valid UTF-8")
                })?;
            }
            4 => {
                require_wire_type(wire_type, 2, "invalid stream frame body wire type")?;
                frame_body = consume_length_delimited(body, &mut offset)?.to_vec();
            }
            5 => return decode_frame::<RpcStreamFrame>(body),
            _ => skip_proto_field(body, &mut offset, wire_type)?,
        }
    }

    if RpcStreamFrameKind::try_from(kind).is_err() {
        return Err(invalid_stream_frame(
            "stream frame contained an unknown frame kind",
        ));
    }

    Ok(RpcStreamFrame {
        kind,
        status,
        message,
        body: frame_body,
        metadata,
    })
}

/// Decodes a protobuf message from a `TrevRPC` frame body.
pub fn decode_frame<M>(body: &[u8]) -> Result<M>
where
    M: Message + Default,
{
    M::decode(body).map_err(Error::from)
}

fn check_frame_body_len(len: usize, max_frame_size: usize) -> Result<()> {
    if len > max_frame_size {
        return Err(Error::FrameTooLarge {
            len,
            max: max_frame_size,
        });
    }

    Ok(())
}

fn frame_len(len: usize, max_frame_size: usize) -> Result<u32> {
    u32::try_from(len).map_err(|_| Error::FrameTooLarge {
        len,
        max: max_frame_size,
    })
}

fn message_stream_frame_body_len(body_len: usize) -> usize {
    if body_len == 0 {
        return 0;
    }

    1 + varint_len_usize(body_len) + body_len
}

fn append_message_stream_frame_body(frame: &mut Vec<u8>, body: &[u8]) {
    if body.is_empty() {
        return;
    }

    frame.push(STREAM_FRAME_BODY_TAG);
    append_varint_usize(frame, body.len());
    frame.extend_from_slice(body);
}

fn append_varint_usize(data: &mut Vec<u8>, mut value: usize) {
    while value >= 0x80 {
        let byte = u8::try_from(value & 0x7f).expect("masked varint byte should fit in u8");
        data.push(byte | 0x80);
        value >>= 7;
    }
    data.push(u8::try_from(value).expect("terminal varint byte should fit in u8"));
}

fn varint_len_usize(mut value: usize) -> usize {
    let mut len = 1;
    while value >= 0x80 {
        len += 1;
        value >>= 7;
    }
    len
}

fn consume_varint(data: &[u8], offset: &mut usize) -> Result<u64> {
    let mut value = 0_u64;
    for shift in (0..64).step_by(7) {
        let Some(byte) = data.get(*offset).copied() else {
            return Err(invalid_stream_frame("truncated stream frame varint"));
        };
        *offset += 1;
        value |= u64::from(byte & 0x7f) << shift;
        if byte < 0x80 {
            return Ok(value);
        }
    }

    Err(invalid_stream_frame("stream frame varint exceeded 64 bits"))
}

fn require_wire_type(actual: u64, expected: u64, message: &'static str) -> Result<()> {
    if actual != expected {
        return Err(invalid_stream_frame(message));
    }

    Ok(())
}

fn consume_length_delimited<'a>(data: &'a [u8], offset: &mut usize) -> Result<&'a [u8]> {
    let len = usize::try_from(consume_varint(data, offset)?)
        .map_err(|_| invalid_stream_frame("stream frame field length exceeded supported range"))?;
    let end = offset
        .checked_add(len)
        .ok_or_else(|| invalid_stream_frame("stream frame field length overflowed"))?;
    if end > data.len() {
        return Err(invalid_stream_frame("truncated stream frame field"));
    }

    let start = *offset;
    *offset = end;
    Ok(&data[start..end])
}

fn skip_proto_field(data: &[u8], offset: &mut usize, wire_type: u64) -> Result<()> {
    match wire_type {
        0 => {
            let _ = consume_varint(data, offset)?;
            Ok(())
        }
        1 => skip_fixed(data, offset, 8),
        2 => {
            let _ = consume_length_delimited(data, offset)?;
            Ok(())
        }
        5 => skip_fixed(data, offset, 4),
        _ => Err(invalid_stream_frame(
            "stream frame contained an unsupported wire type",
        )),
    }
}

fn skip_fixed(data: &[u8], offset: &mut usize, len: usize) -> Result<()> {
    let end = offset
        .checked_add(len)
        .ok_or_else(|| invalid_stream_frame("stream frame fixed field length overflowed"))?;
    if end > data.len() {
        return Err(invalid_stream_frame("truncated stream frame fixed field"));
    }
    *offset = end;
    Ok(())
}

fn invalid_stream_frame(message: &'static str) -> Error {
    Error::from(Status::invalid_argument(message))
}

/// Decodes and validates the body length stored in a `TrevRPC` frame header.
pub fn frame_body_len(header: [u8; FRAME_HEADER_LEN], max_frame_size: usize) -> Result<usize> {
    let len = usize::try_from(u32::from_be_bytes(header)).map_err(|_| Error::FrameTooLarge {
        len: usize::MAX,
        max: max_frame_size,
    })?;

    if len > max_frame_size {
        return Err(Error::FrameTooLarge {
            len,
            max: max_frame_size,
        });
    }

    Ok(len)
}

#[cfg(test)]
mod tests {
    use crate::{Code, RpcRequest, RpcStreamFrame, Status};

    use super::{
        decode_frame, decode_stream_frame_body, encode_frame, encode_message_stream_frame,
        encode_message_stream_frames, frame_body_len,
    };

    #[test]
    fn round_trips_a_frame() {
        let request = RpcRequest::new("hello.Greeter", "SayHello", b"trev".to_vec());
        let frame = encode_frame(&request).expect("request should encode");
        let header = [frame[0], frame[1], frame[2], frame[3]];
        let len = frame_body_len(header, 1024).expect("frame length should decode");
        let decoded = decode_frame::<RpcRequest>(&frame[4..]).expect("request should decode");

        assert_eq!(len, frame.len() - 4);
        assert_eq!(decoded.service, request.service);
        assert_eq!(decoded.method, request.method);
        assert_eq!(decoded.body, request.body);
        assert_eq!(decoded.metadata, request.metadata);
    }

    #[test]
    fn frame_body_len_boundary_cases_are_stable() {
        for length in [0_u32, 1, 15, 16] {
            let decoded =
                frame_body_len(length.to_be_bytes(), 16).expect("in-range length should decode");
            assert_eq!(decoded, length as usize);
        }

        for length in [17_u32, u32::MAX] {
            let error = frame_body_len(length.to_be_bytes(), 16)
                .expect_err("oversized length should be rejected before body allocation");
            assert_eq!(error.into_status().code(), Code::ResourceExhausted);
        }
    }

    #[test]
    fn malformed_protobuf_frame_bodies_are_invalid_argument() {
        for body in [&[0xff][..], &[0xff, 0xff], &[0x0a, 0xff]] {
            let status = decode_frame::<RpcRequest>(body)
                .expect_err("malformed protobuf should fail")
                .into_status();

            assert_eq!(status.code(), Code::InvalidArgument);
        }
    }

    #[test]
    fn stream_message_fast_path_matches_protobuf_encoding() {
        let body = b"hello".to_vec();
        let fast = encode_message_stream_frame(&body, 1024).expect("frame should encode");
        let generic = encode_frame(&RpcStreamFrame::message(body.clone()))
            .expect("generic frame should encode");

        assert_eq!(fast, generic);

        let decoded = decode_stream_frame_body(&fast[4..]).expect("frame should decode");
        assert_eq!(
            decoded.frame_kind(),
            Some(crate::RpcStreamFrameKind::Message)
        );
        assert_eq!(decoded.body, body);
    }

    #[test]
    fn empty_stream_message_fast_path_matches_protobuf_encoding() {
        let fast = encode_message_stream_frame(&[], 1024).expect("frame should encode");
        let generic = encode_frame(&RpcStreamFrame::message(Vec::new()))
            .expect("generic frame should encode");

        assert_eq!(fast, generic);

        let decoded = decode_stream_frame_body(&fast[4..]).expect("frame should decode");
        assert_eq!(
            decoded.frame_kind(),
            Some(crate::RpcStreamFrameKind::Message)
        );
        assert!(decoded.body.is_empty());
    }

    #[test]
    fn batched_stream_message_fast_path_concatenates_frames() {
        let bodies = vec![b"left".to_vec(), b"right".to_vec()];
        let batched = encode_message_stream_frames(&bodies, 1024).expect("frames should encode");
        let mut expected = Vec::new();
        for body in &bodies {
            expected.extend(
                encode_message_stream_frame(body, 1024).expect("single frame should encode"),
            );
        }

        assert_eq!(batched, expected);
    }

    #[test]
    fn stream_frame_fast_decoder_accepts_status_frames() {
        let frame = RpcStreamFrame::status(Status::unavailable("later"));
        let encoded = encode_frame(&frame).expect("status should encode");
        let decoded = decode_stream_frame_body(&encoded[4..]).expect("status should decode");

        assert_eq!(decoded.frame_kind(), frame.frame_kind());
        assert_eq!(decoded.status, frame.status);
        assert_eq!(decoded.message, frame.message);
    }
}
